### DISCUSSION TITLE
bleak pairing-agent branch, bluez backend times out at btmon "@ MGMT Event: User Passkey Request (0x0010) plen 7"

### DISCUSSION THREAD

**Discussion by sigsbeeus** [hbldh/bleak#1348] [Q&A]
Does the pairing-agent branch and bluez backend pair using any method other than "JustWorks"?   I've modified the bluez_agent() function in backends/bluezdbus/agent.py to have the I/O capability "KeyboardDisplay" and the device receiving the pairing request from bluez has the I/O capability "DisplayOnly".  OOB is 0, Security and MITM are 1.  My understanding is this should give me the Authenticated Passkey Entry where bluez should use the Agent RequestPasskey() method to get the passkey.   But the callback doesn't appear to be getting called.   btmon times out after the message:

                       @ MGMT Event: User Passkey Request (0x0010) plen 7  

shows up in the trace, and I never see any debug log messages from either the agent or the agent callbacks.

The same setup works on both Windows and macOS with bleak.  I'm trying to understand why bluez is different.  Has anyone else been successful with this?

Thank you.
--bill

**Comment by dlech** [comment]
What does the Bleak debug logs say?

**Comment by sigsbeeus** [comment]
'll just re-cap and close this. I have two devices I'm developing, one is based on the On Semi RSL10, the other on the STM32WB. I'd like both to be able to use some form of secure bonding across Windows, macOS, Linux, iOS, and Android. I've been trying to get BLE pairing to work by setting the I/O capability to DisplayOnly, and setting the MITM and Secure Connection bits, OOB is zero. The On Semi device's security manager appears to be doing the right thing, and linux looks like it decides to fall back to JustWorks pairing. The STM32WB security manager appears to be implemented either poorly or too rigidly and refuses to pair with linux unless I change the security settings in the device.

On the linux side, it looks like the pairing-agent branch doesn't really work at all. I have yet to see any callback appear to get called out of the bluez stack, even after compiling bluez myself. It's difficult to tell what's going on and I just don't have any more time to track this down.

If someone has an STM32wb-based device that they can use with any pairing method besides JustWorks I'd like to hear how you did it.

Thank you,
--b

**Reply by bojanpotocnik** [reply]
`pairing-agent` branch is outdated (see https://github.com/hbldh/bleak/pull/1100#issuecomment-1337839871) and lacks all fixes added to the latest versions of `bleak`, so it is not surprising if it does not work correctly. I have done some additional work on this topic based on `pairing-agent`, but it is most likely also already outdated and would need another look.

**Comment by sigsbeeus** [comment]
okay, so let me just ask directly:  Does any method besides JustWorks bonding work in any bleak branch for linux/bluez?
